File: src/com/company/Application.java
package com.company;

import java.util.HashMap;
import java.util.UUID;

/**
 * Created by vbaum on 20/09/14.
 */
public class Application {
    private HashMap<Class, HandleCommand> _handlers;

    private ReadModel _db = ReadModel.instance();

    interface HandleCommand {
        void handle(Command cmd) throws Task.TaskAlreadyCreated, Task.TaskYetNotCreated;
    }

    interface HandleEvent<T> {

        void handle(T e);
    }


    public void send(Command command) throws Task.TaskYetNotCreated, Task.TaskAlreadyCreated {

        HandleCommand handler = _handlers.get(command.getClass());
        handler.handle(command);
    }

    public boolean isProcessed(UUID id, int version) {

        TaskItem item = read(id);

//        System.out.println("Expected " + version);

        return item != null && item.version > version;
    }

    public TaskItem read(UUID id) {

        return _db.get(id);
    }

    public Application bootstrap() {

        _handlers = new HashMap<Class, HandleCommand>();

        final TasksHandlers handlers = new TasksHandlers();

        _handlers.put(CreateTask.class, handlers.Create());

        _handlers.put(RenameTask.class, handlers.Rename());

        _handlers.put(CompleteTask.class, handlers.Complete());

        EventBus bus = EventBus.instance().refresh();

        bus.subscribe(new HandleEvent<TaskCreated>() {
            @Override
            public void handle(TaskCreated e) {

                TaskItem item = new TaskItem();
                item.title = e.title;
                item.version = e.version;
                _db.put(e.id, item);
            }
        });

        bus.subscribe(new HandleEvent<TaskRenamed>() {
            @Override
            public void handle(TaskRenamed e) {

                TaskItem item = read(e.id);
                item.title = e.title;
                item.version = e.version;
            }
        });

        bus.subscribe(new HandleEvent<TaskCompleted>() {
            @Override
            public void handle(TaskCompleted e) {

                TaskItem item = read(e.id);
                item.completed = true;
                item.version = e.version;
            }
        });

        return this;
    }
}

File: CommandAndQueries/CommandAndQueries/Application.swift
//
//  Application.swift
//  CommandAndQueries
//
//  Created by Vitaly Baum on 11/10/14.
//  Copyright (c) 2014 Vitaly Baum. All rights reserved.
//

import Foundation

class Application {
    
    let handlers:CommandHandlers;
    
    init (handlers: CommandHandlers) {
        
        self.handlers = handlers

        EventBus.subscribe(self.taskCreated);
        EventBus.subscribe(self.taskRenamed);
        EventBus.subscribe(self.taskCompleted);
    }
    
    var items = Dictionary<String, TaskItem>()
    
    func taskCreated(e: TaskCreated) {
        items[e.id] = TaskItem(title: e.title, version: e.version)
    }
    
    func taskRenamed(e: TaskRenamed) {
        
        let i = read(e.id)
        
        items[e.id] = TaskItem(title: e.title, completed: i.completed, version: e.version)
    }
    
    func taskCompleted(e: TaskCompleted) {
        let i = read(e.id)
        
        items[e.id] = TaskItem(title: i.title, completed: true, version: e.version)
    }
    
    class func create() -> Application {
        
        return Application(handlers: CommandHandlers())
    }
    
    func send(cmd: CreateTask) {
        self.handlers.createTask(cmd);
    }
    
    func send(cmd: RenameTask) {
        self.handlers.renameTask(cmd);
    }
    
    func send(cmd: CompleteTask) {
        self.handlers.completeTask(cmd);
    }
    
    func read(id: String) -> TaskItem {
        let res = items[id] ?? TaskItem(title: "", version: -1)
        
        return res
    }
}
